plugins/dell-dock/fu-dell-dock-usb-usb4.c

```diff
@@ -0,0 +1,654 @@
+						   metadata);
+
+		/* and length field in dwords, note 0 means 16 dwords */
+		metadata[3] = (padded_len / 4) & 0xf;
+
+		/* ask hub to read up to 64 bytes from NVM to mbox data regs */
+		if (!fu_dell_dock_usb4_hub_operation (device, OP_NVM_READ,
+						      metadata, &op_status,
+						      error)) {
+			g_prefix_error (error, "hub firmware NVM read error");
+			return FALSE;
+		}
+		/* FIXME take a look at op_status here */
+
+		/* read the data from mbox data regs into our buffer */
+		if (!fu_dell_dock_usb4_mbox_data_read (device, tmpbuf,
+						       padded_len, error)) {
+			g_prefix_error (error,
+					"hub firmware mbox data read error");
+			return FALSE;
+		}
+		memcpy(buf, tmpbuf + unaligned_bytes, nbytes);
+
+		buf += nbytes;
+		nvm_addr += nbytes;
+		length -= nbytes;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock_usb4_hub_nvm_write (FuDevice *device, const guint8 *buf,
+				 guint32 length, guint32 nvm_addr,
+				 GError **error)
+{
+	guint8 metadata[4];
+	guint8 op_status = 0;
+	guint32 bytes_done = 0;
+	guint32 bytes_total = length;
+
+	if (nvm_addr % 4) {
+		g_prefix_error (error,
+			    "Invalid NVM write offset 0x%x, must be DW aligned",
+			    nvm_addr);
+		return FALSE;
+	}
+	if (length < 64 || length % 64) {
+		g_prefix_error (error,
+			     "Invalid NVM length 0x%x, must be 64 byte aligned",
+			     length);
+		return FALSE;
+	}
+
+	/* 1. Set initial offset, must be DW aligned */
+	fu_dell_dock_usb4_write_u32to_buf (NVM_OFFSET_TO_METADATA(nvm_addr),
+					   metadata);
+
+	if (!fu_dell_dock_usb4_hub_operation (device, OP_NVM_SET_OFFSET,
+					      metadata, &op_status,
+					      error)) {
+		g_prefix_error (error,
+				"hub firmware NVM set offset error, status %d",
+				op_status);
+		return FALSE;
+	}
+
+	/* 2 Write data in 64 byte blocks */
+	fu_device_set_progress_full (device, bytes_done, bytes_total);
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_WRITE);
+	while (length > 0)  {
+		/* write data to mbox data regs */
+		if (!fu_dell_dock_usb4_mbox_data_write (device, buf, 64,
+							error)) {
+			g_prefix_error (error,
+					"hub firmware mbox data write error");
+			return FALSE;
+		}
+		/* ask hub to write 64 bytes from data regs to NVM */
+		if (!fu_dell_dock_usb4_hub_operation (device, OP_NVM_WRITE,
+						      NULL, &op_status, error)) {
+			g_prefix_error (error,
+			    "hub firmware NVM write operation error, status %d",
+			    op_status);
+			return FALSE;
+		}
+		buf += 64;
+		length -= 64;
+		fu_device_set_progress_full (device, bytes_done += 64,
+					     bytes_total);
+	}
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_BUSY);
+	return TRUE;
+}
+
+/* Issue a nvm auth write operation */
+static gboolean
+fu_dell_dock_usb4_hub_nvm_authenticate (FuDevice *device, GError **error)
+{
+	guint8 op_status;
+
+	if (!fu_dell_dock_usb4_hub_operation (device, OP_NVM_AUTH_WRITE,
+					       NULL, &op_status, error)) {
+		g_prefix_error (error,
```

Comment:
how common is this?  Should we worry about this coming up to the tools as `FALSE`?  I wonder if instead you should return TRUE in this instance and just output the message into the daemon logs with `g_print` or something?